Monte carlo Improvement

GbotemiB opened the pull request:
# Closes # (if applicable).

## Changes proposed in this Pull Request
This PR includes improvement on the Monte-Carlo simulation. The improvement includes the option to choose from different distributions for the simulation. The distribution includes normal, triangle, lognormal, beta, and gamma. While choosing the distribution, it also allows users to pass parameters for the distribution.

## Checklist

- [x] I consent to the release of this PR's code under the AGPLv3 license and non-code contributions under CC0-1.0 and CC-BY-4.0.
- [x] I tested my contribution locally and it works fine.
- [ ] Code and workflow changes are sufficiently documented.
- [ ] Newly introduced dependencies are added to `envs/environment.yaml` and `doc/requirements.txt`.
- [x] Changes in configuration options are added in all of `config.default.yaml` and `config.tutorial.yaml`.
- [ ] Add a test config or line additions to `test/` (note tests are changing the config.tutorial.yaml)
- [ ] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [ ] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

davide-f commented on scripts/monte_carlo.py:
```diff
@@ -147,20 +192,145 @@ def monte_carlo_sampling_scipy(
 
 
+def rescale_distribution(
+    latin_hypercube: np.array, distribution: str, distribution_params: list
```
currently we don't have the hinting. It would be interesting to discuss on whether we should implement.
The discussion may extend to pypsa-eur as well

GbotemiB replied:
oh...I can move this conversation to the discord channel

GbotemiB replied:
I agree to initiate the discussion on this.

davide-f replied:
That could be nice to discuss in the pypsa meets earth channel, ideally also during the pypsa-earth periodic meetings :)
That's supernice to hear the perspective of CS :)

GbotemiB replied:
Awesome

davide-f replied:
moved to the discussion :) resolving here

GbotemiB commented on scripts/monte_carlo.py:
```diff
@@ -14,15 +14,30 @@
-        add_to_snakefile:
-        samples:
-        sampling_strategy:
+        add_to_snakefile: false
```
Regarding this, I think we can leave it with relative value, then we can set an option for scaling to be relative or absolute.

davide-f replied:
agree

GbotemiB commented on scripts/monte_carlo.py:
```diff
@@ -192,13 +362,21 @@ def monte_carlo_sampling_scipy(
 
     ### SCENARIO CREATION / SAMPLING STRATEGY
     ###
     if SAMPLING_STRATEGY == "pydoe2":
```
I have restructured the implementation. I am currently testing it.

davide-f commented on config.tutorial.yaml:
```diff
@@ -337,18 +337,39 @@ costs:
+    sampling_strategy: "chaospy"  # "pydoe2", "chaospy", "scipy", packages that are supported
+  uncertainties:
+    loads_t.p_set:
+      scale: [0.9, 1.1]
+      type: Uniform
+      args: [0, 1]
```
Nice proposal, as a comment:
to me, the use of scale and args, seem to be a duplication in the definition of uncertainties.
For example, for uniform, we could use directly:  args: [0.9, 1.1], or even explicitly state: lower_bound: 0.9, upper_bound: 1.1 (keys matching the arguments of the scipy function for example).
What do you think?

GbotemiB replied:
True. Thats a clear point. Thank you very much. Args is meant to be the parameters for the chosen distribution.

davide-f replied:
Mmm not sure if it is just a naming convention.
If we dropped entirely the scale factor and use args: [0.9, 1.1], I'd expect the uniform distribution to be created between 0.9 and 1.1.
A similar behavior may be expected for the other distributions.
The idea is that the arguments already describe the values of the distribution.

The concept of bounds is different and may apply for some continuous distributions (e.g. gaussian) when we may want to truncate some values, but the concepts are different and I'm unsure we want to delve into that.

By looking at the code, probably the scale is kind-of due to historical reasons but may be now revised.
For each scenario, we always first draw a vector of uniform distributions between 0 and 1 for every uncertain parameter, then we convert the uniform sampling into the desired distribution using the ppf methodology.
The scale factor is no more needed as the first extraction is always between 0 and 1 and then, the uniform sampling is mapped into the desired distribution using the inverse function.
The arguments (args) of each distribution already account for the shape of the desired distribution.
Does it make sense?

GbotemiB replied:
I think args is particularly useful to vary the shape of the desired distribution of the user when the desired chosen distribution is either gamma, beta, lognormal, normal, or Triangular. 

The goal here is to give users the flexibility to control the shape of selected distributions that aren't uniform.

davide-f replied:
the standard uniform yes, but in our application the lower and upper bounds may change as discussed and reviewed.
Resolving this conversation

GbotemiB commented on scripts/monte_carlo.py:
```diff
@@ -147,20 +192,145 @@ def monte_carlo_sampling_scipy(
+        latin_hypercube = beta.ppf(latin_hypercube, a, b)
+    elif distribution == "Gamma":
+        shape, scale = distribution_params
+        latin_hypercube = gamma.ppf(latin_hypercube, shape, scale)
```
Just a bit of clarity here, are you referring to python structural matching here or regex matching?

davide-f replied:
Solved in a past pypsa-earth meeting

GbotemiB commented:
Hello @davide-f, I have implemented all the requested changes. I think the PR is ready for another round of review.

davide-f commented on scripts/monte_carlo.py:
```diff
@@ -147,20 +189,161 @@ def monte_carlo_sampling_scipy(
+        params = value.get("args")
+
+        match dist:
+            case "Uniform":
```
The uniform distribution may have different lower and upper bounds, maybe better to do a scaling (that is easy do)

GbotemiB replied:
That could work but eventually, the result will rescaled between 0 and 1 using the minmax scaler.

davide-f replied:
ok, so maybe we could implement something that with the fit_trasform turns out to be what we desired.
Not sure if you already implemented that

GbotemiB replied:
Yeah. Would fix this too

davide-f commented on scripts/monte_carlo.py:
```diff
@@ -147,20 +189,161 @@ def monte_carlo_sampling_scipy(
+                latin_hypercube[:, idx] = norm.ppf(latin_hypercube[:, idx], mean, std)
+            case "LogNormal":
+                mean, std = params
+                latin_hypercube[:, idx] = lognorm.ppf(latin_hypercube[:, idx], s=0.90)
```
what is s=0.9?
should that be another parameter?

GbotemiB replied:
Yeah. I should pass it through the parameters var

davide-f replied:
yeah, isn't it "shape"?

GbotemiB replied:
You are correct. It is shape.

davide-f commented on config.tutorial.yaml:
```diff
@@ -337,18 +337,36 @@ costs:
 
 
 monte_carlo:
```
For reproducibility, it may be good to add an optional option on the seed, like "seed: 42".

GbotemiB replied:
Absolutely

davide-f reviewed (commented):
Great revision @GbotemiB :D
Few minor comments :)

As a comments, in debug, may you show a plot of input and outputs provided by the rescaling function for different values?
Have you done that to make sure the expected distributions are verified?

davide-f commented:
@GbotemiB Emmanuel, please use precommit.
in github a ton of pre-commit issues are appearing, please check. It would be good to clean a bit the history by resetting to some past commit like [this one](https://github.com/pypsa-meets-earth/pypsa-earth/pull/930/commits/e79566a271a9d3fd11d3035b33dc5e987129a827) and placing the changes.

To make sure you use the pre-commit, please check [this](https://pypsa-earth.readthedocs.io/en/latest/how_to_contribute.html#linting-and-pre-commit) that suggests you how to set it up

davide-f commented on scripts/monte_carlo.py:
```diff
@@ -180,47 +363,49 @@ def monte_carlo_sampling_scipy(
     ###
     MONTE_CARLO_PYPSA_FEATURES = {
-        k: v for k, v in monte_carlo_config["pypsa_standard"].items() if v
+        k: v.get("scale") for k, v in monte_carlo_config["uncertainties"].items() if v
```
do we have scale?
I think we have same legacy to fix here

GbotemiB replied:
Haa...No we no longer have scale.

davide-f reviewed (commented):
Partial revision for few comments :D

GbotemiB commented:
Hi @davide-f, I have successfully cleaned up the history. Thanks for the tip.

Also, I have made changes to the part of the code you reviewed. Please feel free to review the changes. Thanks.

GbotemiB commented:
> Great revision @GbotemiB :D Few minor comments :)
> 
> As a comments, in debug, may you show a plot of input and outputs provided by the rescaling function for different values? Have you done that to make sure the expected distributions are verified?

I have tested that the rescaling works. Here is what the rescaled distribution looks like using 1369 samples
<img width="1500" alt="image" src="https://github.com/pypsa-meets-earth/pypsa-earth/assets/48842684/766dbb98-03d8-43d9-ab94-21b8487d331e">

GbotemiB commented:
Hello @davide-f, I added into the `monte-carlo.py` script a few lines to plot the selected distribution. I also added a rule to the snakemake file to separate the cleaning process for `solve_all_networks_monte` rule.

Happy to hear your reviews 🙃

davide-f commented on Snakefile:
```diff
@@ -65,16 +65,24 @@ if config["custom_rules"] is not []:
         shell("snakemake -j 1 run_all_scenarios --delete-all-output")
 
 
+rule clean:
```
The original clean also cleaned the monte carlo.
Do you think a new rule is needed?

GbotemiB replied:
```bash
rule clean:
    run:
        shell("snakemake -j 1 solve_all_networks --delete-all-output")
        try:
            shell("snakemake -j 1 solve_all_networks_monte --delete-all-output")
        except:
            pass
        shell("snakemake -j 1 run_all_scenarios --delete-all-output")
```
This is the original clean. The first `snakemake` doesn't get executed.
Here is the error log in the terminal
```
Restricted license - for non-production use only - expires 2025-11-24
/Users/gbotemi/miniconda3/envs/pypsa-earth-v2/lib/python3.10/site-packages/pulp/apis/gurobi_api.py:170: UserWarning: GUROBI error: Unknown parameter '_test'.
  warnings.warn("GUROBI error: {}.".format(e))
/Users/gbotemi/miniconda3/envs/pypsa-earth-v2/lib/python3.10/site-packages/pypsa/networkclustering.py:16: UserWarning: The namespace `pypsa.networkclustering` is deprecated and will be removed in PyPSA v0.24. Please use `pypsa.clustering.spatial instead`. 
  warnings.warn(
Building DAG of jobs...
WARNING:snakemake.logging:Building DAG of jobs...
Using shell: /bin/bash
WARNING:snakemake.logging:Using shell: /bin/bash
Provided cores: 1 (use --cores to define parallelism)
WARNING:snakemake.logging:Provided cores: 1 (use --cores to define parallelism)
Rules claiming more threads will be scaled down.
WARNING:snakemake.logging:Rules claiming more threads will be scaled down.
Job stats:
job      count
-----  -------
clean        1
total        1

WARNING:snakemake.logging:Job stats:
job      count
-----  -------
clean        1
total        1

Select jobs to execute...
WARNING:snakemake.logging:Select jobs to execute...

INFO:snakemake.logging:
[Fri Jan 12 11:26:39 2024]
INFO:snakemake.logging:[Fri Jan 12 11:26:39 2024]
rule clean:
    jobid: 0
    reason: Rules with neither input nor output files are always executed.
    resources: tmpdir=/var/folders/c3/657p__892k72qxh6r8fdkqsh0000gn/T
INFO:snakemake.logging:rule clean:
    jobid: 0
    reason: Rules with neither input nor output files are always executed.
    resources: tmpdir=/var/folders/c3/657p__892k72qxh6r8fdkqsh0000gn/T

INFO:snakemake.logging:
Restricted license - for non-production use only - expires 2025-11-24
/Users/gbotemi/miniconda3/envs/pypsa-earth-v2/lib/python3.10/site-packages/pulp/apis/gurobi_api.py:170: UserWarning: GUROBI error: Unknown parameter '_test'.
  warnings.warn("GUROBI error: {}.".format(e))
/Users/gbotemi/miniconda3/envs/pypsa-earth-v2/lib/python3.10/site-packages/pypsa/networkclustering.py:16: UserWarning: The namespace `pypsa.networkclustering` is deprecated and will be removed in PyPSA v0.24. Please use `pypsa.clustering.spatial instead`. 
  warnings.warn(
Building DAG of jobs...
WARNING:snakemake.logging:Building DAG of jobs...
Using shell: /bin/bash
WARNING:snakemake.logging:Using shell: /bin/bash
Provided cores: 1 (use --cores to define parallelism)
WARNING:snakemake.logging:Provided cores: 1 (use --cores to define parallelism)
Rules claiming more threads will be scaled down.
WARNING:snakemake.logging:Rules claiming more threads will be scaled down.
Select jobs to execute...
WARNING:snakemake.logging:Select jobs to execute...
Restricted license - for non-production use only - expires 2025-11-24
/Users/gbotemi/miniconda3/envs/pypsa-earth-v2/lib/python3.10/site-packages/pulp/apis/gurobi_api.py:170: UserWarning: GUROBI error: Unknown parameter '_test'.
  warnings.warn("GUROBI error: {}.".format(e))
/Users/gbotemi/miniconda3/envs/pypsa-earth-v2/lib/python3.10/site-packages/pypsa/networkclustering.py:16: UserWarning: The namespace `pypsa.networkclustering` is deprecated and will be removed in PyPSA v0.24. Please use `pypsa.clustering.spatial instead`. 
  warnings.warn(
Building DAG of jobs...
WARNING:snakemake.logging:Building DAG of jobs...
MissingInputException in rule solve_all_networks in file /Users/gbotemi/Documents/code/PYPSA/mc/pypsa-earth/Snakefile, line 102:
Missing input files for rule solve_all_networks:
    affected files:
        results/NG/networks/elec_s_10_ec_lcopt_Co2L-24H.nc
ERROR:snakemake.logging:MissingInputException in rule solve_all_networks in file /Users/gbotemi/Documents/code/PYPSA/mc/pypsa-earth/Snakefile, line 102:
Missing input files for rule solve_all_networks:
    affected files:
        results/NG/networks/elec_s_10_ec_lcopt_Co2L-24H.nc
[Fri Jan 12 11:26:43 2024]
INFO:snakemake.logging:[Fri Jan 12 11:26:43 2024]
Error in rule clean:
    jobid: 0

ERROR:snakemake.logging:Error in rule clean:
    jobid: 0

RuleException:
CalledProcessError in file /Users/gbotemi/Documents/code/PYPSA/mc/pypsa-earth/Snakefile, line 69:
Command 'set -euo pipefail;  snakemake -j 1 solve_all_networks --delete-all-output' returned non-zero exit status 1.
  File "/Users/gbotemi/Documents/code/PYPSA/mc/pypsa-earth/Snakefile", line 69, in __rule_clean
  File "/Users/gbotemi/miniconda3/envs/pypsa-earth-v2/lib/python3.10/concurrent/futures/thread.py", line 58, in run
ERROR:snakemake.logging:RuleException:
CalledProcessError in file /Users/gbotemi/Documents/code/PYPSA/mc/pypsa-earth/Snakefile, line 69:
Command 'set -euo pipefail;  snakemake -j 1 solve_all_networks --delete-all-output' returned non-zero exit status 1.
  File "/Users/gbotemi/Documents/code/PYPSA/mc/pypsa-earth/Snakefile", line 69, in __rule_clean
  File "/Users/gbotemi/miniconda3/envs/pypsa-earth-v2/lib/python3.10/concurrent/futures/thread.py", line 58, in run
Shutting down, this might take some time.
WARNING:snakemake.logging:Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
ERROR:snakemake.logging:Exiting because a job execution failed. Look above for error message
Shutting down, this might take some time.
WARNING:snakemake.logging:Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
ERROR:snakemake.logging:Exiting because a job execution failed. Look above for error message
Complete log: .snakemake/log/2024-01-12T112636.677565.snakemake.log
WARNING:snakemake.logging:Complete log: .snakemake/log/2024-01-12T112636.677565.snakemake.log
```

GbotemiB replied:
We might not need a new rule. Modifying the existing rule should be sufficient.

davide-f replied:
Mmmm interesting, may you try removing the -j 1 in the commands? that may solve the error

GbotemiB replied:
I tried that. It didn't work. The -j 1 command is required at the very least to make the command work.

GbotemiB replied:
I have modified the clean rule. The clean rule now works for cleaning monte-carlo.

davide-f commented on config.default.yaml:
```diff
@@ -25,7 +25,7 @@ enable:
 
 run:
-  name: "" # use this to keep track of runs with different settings
+  name: "NG_BJ" # use this to keep track of runs with different settings
```
Please revert

GbotemiB replied:
Thanks for the observation. I have reverted this in the recent commits.

davide-f commented on config.default.yaml:
```diff
@@ -340,18 +340,38 @@ costs:
+    # as key: add below the pypsa object for the monte_carlo sampling, "network" is only allowed for filtering!
+    # as value: currently supported format [l_bound, u_bound] or empty [], represent multiplication factors for the object
   options:
     add_to_snakefile: false
```
Please add a comment to explicit that when true, monte carlo is enabled

GbotemiB replied:
Thanks for the observation. I have fixed it in the recent commits.

davide-f commented on config.default.yaml:
```diff
@@ -340,18 +340,38 @@ costs:
 
 
 monte_carlo:
+    # uniform: https://chaospy.readthedocs.io/en/master/api/chaospy.Uniform.html
```
This description is unfortunately not clear.

I drop here few comments to potentially improve:
- We could add a first line describing the option list as a short introduction. we should mention that a different distribution can be specified to any different pypsa network parameter and the supported distributions are the following
- I think we can avoid a link for every option, we could reference the chaospy reference guide and mention the supported distributions. 
- the details on how to specify the parameters for each distribution may be moved below under "uncertainties". There it is worth adding few comments to specify how to specify the parameters. now it is kind-of everything at the beginning and it may be a bit complex

GbotemiB replied:
Thanks for the review. I have made changes to the descriptions.

davide-f commented on scripts/monte_carlo.py:
```diff
@@ -177,71 +370,79 @@ def monte_carlo_sampling_scipy(
-    lh_scaled = qmc.scale(lh, L_BOUNDS, U_BOUNDS)
 
-    ### MONTE-CARLO MODIFICATIONS
+    # create plot for the rescaled distributions
+    for idx in range(N_FEATURES):
+        sns.displot(lh[:, idx], kde=True).set(
+            title=f"{MONTE_CARLO_PYPSA_FEATURES[idx]}"
+        ).figure.savefig(f"{MONTE_CARLO_PYPSA_FEATURES[idx]}.png", bbox_inches="tight")
```
This is a debugging code that I don't dislike to keep as it may be useful for development purposes, but I'd keep it commented.
Accordingly, I'd revice the title to highlight that, e.g. "# create plot for the rescaled distributions (for development usage, commented by default)"

GbotemiB replied:
I have commented this part out in the recent commits.

davide-f reviewed (commented):
Great @GbotemiB :D
Added few comments :)
Can you also post here the images that you created with the monte carlo? very curious and useful to review!

GbotemiB commented:
> Great @GbotemiB :D Added few comments :) Can you also post here the images that you created with the monte carlo? very curious and useful to review!

This is another simulation with 49 samples.

Here is the plot for the objective values for the 49 networks.
![image](https://github.com/pypsa-meets-earth/pypsa-earth/assets/48842684/6a927401-0578-4cfe-be18-539256a5c064)

Here are the distribution plots for the parameters.
![image](https://github.com/pypsa-meets-earth/pypsa-earth/assets/48842684/b40a8748-af38-40aa-9e92-fc760af91317)
![image](https://github.com/pypsa-meets-earth/pypsa-earth/assets/48842684/6b0b0130-d2a1-4f11-9db1-e450ea8bcb23)
![image](https://github.com/pypsa-meets-earth/pypsa-earth/assets/48842684/cb25bd4c-6286-4ed3-a0bf-ab83b2e16c7e)
![image](https://github.com/pypsa-meets-earth/pypsa-earth/assets/48842684/e069bf9f-ae47-404a-ae25-62332f688a0e)
![image](https://github.com/pypsa-meets-earth/pypsa-earth/assets/48842684/1985949e-4a98-4b98-8eb6-3907009f98bf)
![image](https://github.com/pypsa-meets-earth/pypsa-earth/assets/48842684/cf60e7f2-4f3d-4cd0-bb86-f192f891c8be)
![image](https://github.com/pypsa-meets-earth/pypsa-earth/assets/48842684/bc2f6b1a-9e2e-40cb-80c9-659b9f3e5b3a)

davide-f commented on config.default.yaml:
```diff
@@ -340,18 +340,34 @@ costs:
+
+  # Different distributions can be specified for various PyPSA network parameters.
+  # Supported distributions for uncertainties are uniform, normal, lognormal, triangle, beta and gamma.
+  # More info on the distributions are documented in the Chaospy reference guide...
+  # https://chaospy.readthedocs.io/en/master/reference/distribution/index.html
+  # ... users can add flexibly more features for the Monte-Carlo sampling
```
it may be good to specify here how to clarify the parameters, in particular the following:
    {pypsa network object, e.g. "loads_t.p_set"}:
      type: {any distribution among: "uniform", "gaussian", ...}
      args: {arguments passed as a list depending on the distribution, see below supported ones}

GbotemiB replied:
Thanks for the clarity @davide-f. I have included in the comments section more comments clarifying the parameters as suggested.

Feel free to review.

davide-f reviewed (commented):
Nice @GbotemiB !
I was curious in validating the extraction of the coefficients by the monte-carlo.
If you have the network basic elec network for example, you could re-obtain the coefficient for example doing:
coeff = sum(n_montecarlo.generators_t.p_max_pu[:, ... == solar ...]) / sum(n_elecnetwork.generators_t.p_max_pu[:, ... == solar ...])

the example above is for solar, but you could do the same for the other carriers and try get the curves.
I'd like to validate that what we expect being a gaussian is a gaussian for example.

I believe we can soon merge the PR :)

GbotemiB commented:
Hi @davide-f, I can confirm that the post-processed coefficients follow the pre-processed coefficients.
![image](https://github.com/pypsa-meets-earth/pypsa-earth/assets/48842684/97d11f3d-9009-4521-b378-43fe6926da02)
<img width="834" alt="image" src="https://github.com/pypsa-meets-earth/pypsa-earth/assets/48842684/de021338-2d74-4cdc-8fd2-4f08aa7b6d54">

These are some of the plots generated. I will include the plots in the notebook I am working on for the case study analysis and share the notebook with you.

davide-f commented:
Hello @GbotemiB ,
Thanks for the images, but few comments:
- I expected a uniform distribution for the load, why does it have a different shape instead?
- Similarly, for the onwind I expected a lognormal shape but this seems to be normal instead, while for sigma=1.5 I expected a different shape.

Could you crosscheck?

davide-f commented on config.default.yaml:
```diff
@@ -340,18 +340,35 @@ costs:
+  # ... users can add flexibly more features for the Monte-Carlo sampling using the description below
+  # {pypsa network object, e.g. "loads_t.p_set"}:
+  # type: {any distribution among: "uniform", "normal", "lognormal", "triangle", "beta" and "gamma"}
+  # args: {arguments passed as a list depending on the distribution, see arguments description in Chaospy reference guide}
```
There is the need to clarify the order of the arguments.
I see three options:
- add it into the readthedocs description and a crosslink here (if the docstring of the monte_carlo.py file si done well that works)
- add it here
- describe it into the examples below under uncertainties

Point 1 is the best; actually, it should be good to add the monte_carlo here into the API reference at least: https://github.com/pypsa-meets-earth/pypsa-earth/blob/main/doc/api_reference.rst

Here adding some description of at least few rules would be good too, at least doing point 3.
Feel free to propose something that looks compact to you :)

GbotemiB replied:
@davide-f, Thanks for the suggestions. I have planned to include the instructions in the `readthedocs`

davide-f reviewed (commented):
Hello, closing to finalize, I added a comment on the shapes of the distribution that do not look like expected.
Could you clarify?

GbotemiB commented:
> Hello @GbotemiB , Thanks for the images, but few comments:
> 
> * I expected a uniform distribution for the load, why does it have a different shape instead?
> * Similarly, for the onwind I expected a lognormal shape but this seems to be normal instead, while for sigma=1.5 I expected a different shape.
> 
> Could you crosscheck?

Regarding the distribution of the parameters, it follows this config. But I could reconfigure the parameters.
```python
uncertainties:
    loads_t.p_set:
      type: beta
      args: [2, 0.5]
    generators_t.p_max_pu.loc[:, n.generators.carrier == "solar"]:
      type: beta
      args: [4, 1]
    generators_t.p_max_pu.loc[:, n.generators.carrier == "onwind"]:
      type: beta
      args: [4, 1]
    generators.capital_cost.loc[n.generators.carrier == "solar"]:
      type: normal
      args: [0, 1]
    generators.capital_cost.loc[n.generators.carrier == "onwind"]:
      type: normal
      args: [0, 1]
    generators.weight.loc[n.generators.carrier == "onwind"]:
      type: beta
      args: [2, 0.5]
    generators.weight.loc[n.generators.carrier == "solar"]:
      type: beta
      args: [2, 0.5]
```

davide-f commented on scripts/monte_carlo.py:
```diff
@@ -13,19 +13,37 @@
 .. code:: yaml
 
     monte_carlo:
```
You can safely delete the comments to the settings here, as for the other scripts

GbotemiB replied:
I have removed the comments as advised.

davide-f commented on config.default.yaml:
```diff
@@ -340,18 +340,37 @@ costs:
+    samples: 9 # number of optimizations. Note that number of samples when using scipy has to be the square of a prime number
+    sampling_strategy: "chaospy"  # "pydoe2", "chaospy", "scipy", packages that are supported
+    seed: 42 # set seedling for reproducibilty
+  # Different distributions can be specified for various PyPSA network object.
+  # Supported distributions for uncertainties are uniform, normal, lognormal, triangle, beta and gamma.
+  # normal: [mean, std], lognormal: [mean, std], uniform: [lower_bound, upper_bound],
+  # triangle: [mid_point (between 0 - 1)], beta: [alpha, beta], gamma: [shape, scale]
+  # More info on the distributions are documented in the Chaospy reference guide...
+  # https://chaospy.readthedocs.io/en/master/reference/distribution/index.html
+  # ... users can add flexibly more features for the Monte-Carlo sampling using the description below
+  # {pypsa network object, e.g. "loads_t.p_set"}:
+  # type: {any distribution among: "uniform", "normal", "lognormal", "triangle", "beta" and "gamma"}
+  # args: {arguments passed as a list depending on the distribution, see arguments description in Chaospy reference guide}
```
Maybe some reordering? they were a bit disconnected. Feel free to improve :)
```suggestion
  # Uncertanties on any PyPSA object are specified by declaring the specific PyPSA object under the key 'uncertainties'.
  # For each PyPSA object, the 'type' and 'args' keys represent the type of distribution and its argument, respectively.
  # Supported distributions types are uniform, normal, lognormal, triangle, beta and gamma.
  # The arguments of the distribution are passed using the key 'args'  as follows, tailored by distribution type
  # normal: [mean, std], lognormal: [mean, std], uniform: [lower_bound, upper_bound],
  # triangle: [mid_point (between 0 - 1)], beta: [alpha, beta], gamma: [shape, scale]
  # More info on the distributions are documented in the Chaospy reference guide...
  # https://chaospy.readthedocs.io/en/master/reference/distribution/index.html
  # An abstract example is as follows:
  # {pypsa network object, e.g. "loads_t.p_set"}:
  # type: {any supported distribution among the previous: "uniform", "normal", ...}
  # args: {arguments passed as a list depending on the distribution, see the above and more at https://pypsa.readthedocs.io/}
```

davide-f reviewed (commented):
Final rounds!
Great you added the reference to the api :) close to done! [I know it's the third time, but we are really there :)]
Happy to jump in if necessary.

Regarding the distribution plots, have you tested all distributions?

GbotemiB commented:
> Final rounds! Great you added the reference to the api :) close to done! [I know it's the third time, but we are there :)] Happy to jump in if necessary.
> 
> Regarding the distribution plots, have you tested all distributions?

@davide-f, No problem about continuing the review if necessary. 
Regarding the distribution, I will reconfirm all distribution again. I made a few changes to the triangle distribution.

Feel free to jump in.

GbotemiB commented:
@davide-f, Regarding the distribution, I have tested them. Here is a config for Monte-Carlo and the results.

```python
monte_carlo:
  options:
    add_to_snakefile: true
    samples: 961 # number of optimizations. Note that number of samples when using scipy has to be the square of a prime number
    sampling_strategy: "pydoe2"  # "pydoe2", "chaospy", "scipy", packages that are supported
    seed: 42 # set seedling for reproducibilty
  uncertainties:
    loads_t.p_set:
      type: normal
      args: [1, 100]
    generators_t.p_max_pu.loc[:, n.generators.carrier == "solar"]:
      type: uniform
      args: [0.1, 1]
    generators_t.p_max_pu.loc[:, n.generators.carrier == "onwind"]:
      type: lognormal
      args: [1.5]
    generators.capital_cost.loc[n.generators.carrier == "solar"]:
      type: triangle
      args: [0.7]
    generators.capital_cost.loc[n.generators.carrier == "onwind"]:
      type: beta
      args: [0.5, 1]
    generators.weight.loc[n.generators.carrier == "onwind"]:
      type: beta
      args: [2, 0.5]
    generators.weight.loc[n.generators.carrier == "solar"]:
      type: gamma
      args: [2, 0.5]
```

Here are the plots.
![image](https://github.com/pypsa-meets-earth/pypsa-earth/assets/48842684/8aab2582-51a2-45ba-bca8-7b102073e33a)
![image](https://github.com/pypsa-meets-earth/pypsa-earth/assets/48842684/adbedeaa-8745-45bc-9722-52b76596c59a)
![image](https://github.com/pypsa-meets-earth/pypsa-earth/assets/48842684/355df940-9306-4b1d-aa90-8dc551f51237)
![image](https://github.com/pypsa-meets-earth/pypsa-earth/assets/48842684/77004448-fb7b-4708-b4d3-e294e0b6602f)
![image](https://github.com/pypsa-meets-earth/pypsa-earth/assets/48842684/1f352164-f710-42a5-9aa2-98d187dc0961)
![image](https://github.com/pypsa-meets-earth/pypsa-earth/assets/48842684/200e5241-00a5-431d-8552-9bb9692fbeb9)
![image](https://github.com/pypsa-meets-earth/pypsa-earth/assets/48842684/4b5b2db6-9f57-4e17-8bed-5c4ca43063a0)

davide-f reviewed (approved):
Great @GbotemiB :D
The PR is ready to merge, really great great job :D

GbotemiB commented:
Thank you for the support @davide-f. Special shoutout to @ekatef 🎉 🥳

ekatef commented:
Fantastic news, @GbotemiB and @davide-f! A really cool feature to have 😄 
Congratulations with an amazing result! 🎉 🎉 🎉
